### PR TITLE
Rollup adding support for date field metrics (#34185)

### LINKED ISSUE #34185: Add `date` value type support for certain metrics in Rollup
Currently, when creating a new `RollupJobConfig` only numeric field mapper types are supported for the different metrics. 

Support for `min`, `max` and, `value_count` over `date` fields should be added. Since `date` can just be treated as a `long`, hopefully the change is not too pervasive. 

This issue is blocking ML support for rollup indices: #31999 

Pinging @colings86 @polyfractal

### PR DESCRIPTION
**Pull request by benwtrent**
This allows `date` data field types to be rolled up with our supported metric aggregations. 

closes #34185

### CODE REVIEW

**Review by colings86 — CHANGES_REQUESTED** [review]
I think we are going to want to restrict the supported metrics to `min`, `max, and `value_count` here. The reason is that apart from `sum` and `avg` not making much sense, it would take less than 10,000,000 rolled up documents to overflow a long with dates which would probably lead to some very strange behaviours. I can see people reaching 10,000,000 original documents in a single rolled up document but more than that I can definitely see this overflow happening on the query side.

**Comment by benwtrent** [comment]
@colings86 Think I addressed your issue, just waiting on @polyfractal for the final stamp of approval :)

**Review by colings86 — APPROVED** [review]
I left a nit comment but otherwise LGTM. I would like @polyfractal to give it the green light before this is merged though

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/RollupField.java
Comment (polyfractal) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/RollupField.java]: I wonder if we should have `SUPPORTED_NUMERIC_METRICS`, `SUPPORTED_DATE_METRICS`, and then a union of those in a third list `SUPPORTED_METRICS`?

E.g. in the MetricsConfig constructor (and a few other places), we validate that an agg is one of the supported metrics.  We don't yet have the field mapper type, so we can't validate numeric vs date field and their associated metrics, but it's a quick check to make sure the config is at least sane.

The current `SUPPORTED_METRICS` is technically a superset of the date metrics too so everything is fine.  But I'm worried in the future the two lists may diverge for whatever reason, and we miss the validation because dates support something numerics don't, or vice versa.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/MetricConfig.java
Comment (polyfractal) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/MetricConfig.java]: Now that there are two cases, I wonder if we should push the isAggregatable() check out of the type conditionals (e.g. apply to everything), then have an `if...if else...else` for the types?  That way we won't need another conditional inside the first that specializes for the date type.

Not sure how that would look, so the current way is fine if that ends up being messier. :)

Has a nice side effect that the validation error will include both a message about non-aggregatable field as well as the metric missing at the same time.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/MetricConfig.java
Comment (polyfractal) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/MetricConfig.java]: Super tiny nit: could we wrap the `SUPPORTED_DATE_METRICS` with brackets (`[...]`)?  Makes it easier for the user to see the difference in message and dynamic values.

Ditto to below with `unsupportedMetrics`

**Review by polyfractal — CHANGES_REQUESTED** [review]
Left a few questions, minor tweaks.

~Also, I think we need to add a test to `RollupRequestTranslationTests` too.  I believe the translation will "work" (not throw an error), but won't translate correctly due to field naming conventions.~

~Timestamps are stored as `foo.date_histogram.timestamp` + `foo.date_histogram.interval` + `foo.date_histogram.time_zone`.  Metric values, however, are stored as `foo.max.value`.~ 

~Which means [we treat all metrics as defaulting to `.value` when rewriting](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/RollupRequestTranslator.java#L495), and that won't match the `.timestamp` that date_histos use.  I think that will break the rewriting, and we need to figure out a way to change from `.value` to `.timestamp`.  I haven't looked to closely yet, but we could probably do something like detect if the agg is targeting the string `"date_histogram"` (since the convention is defined) and switch on that.~

~Probably good to verify on the response translation side too... but I think that side should "just work" once the request is rewritten correctly.~

~*And yes, it would have been better if date_histos just used `.value` for their timestamp.  alas :(*~

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/MetricConfig.java
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/job/MetricConfig.java]: @polyfractal A `Collection<String>` object when transformed to a `String` transforms to a `[Metric1, Metric2]`. To clarify, you want me to add another bracket so that the printout will be `[[Metric1, Metric2]]` ?

Reference Repl: https://repl.it/@ben_w_trent/QuerulousSubtleApplicationframework

**Comment by polyfractal** [comment]
Chatted on Slack, the rewriting `.timestamp` business doesn't apply (I confused myself).  We will be adding distinct min/max metrics on the timestamp field, rather than reusing the `.timestamp` field that the date_histo generates.

Introduces a bit of duplication in the document but keeps everything simple.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/RollupField.java
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/rollup/RollupField.java]: So, I am making `SUPPORTED_NUMERIC_METRICS` and `SUPPORTED_DATE_METRICS` two separate lists with overlapping metrics. 

Then there is a `Set<String>` that will be `SUPPORTED_METRICS` that is a union of the two (need to verify that the type change will not break other logic). This should lay the ground work/pattern for future non-overlapping supported metrics.

**Review by polyfractal — APPROVED** [review]

**Comment by polyfractal** [comment]
LGTM thanks! <3

**Pull request merged**
